iterators.cpp

```diff
@@ -0,0 +1,66 @@
+#include <iostream>
+#include <vector>
+#include <iterator>
+
+
+//reverse iteration
+
+int reverse_order() {
+
+    std::vector<int> v = {1, 2, 3, 4};
+    auto it = v.rbegin();
+    auto start = v.rend();
+
+    for(; it != start; it++) {
+        std::cout << *it << std::endl;
+    }
+
+    return 0;
+}
+
+
+// iterators for string
+
+int wordy() {
+    std::string s = "Ala ma kota";
+
+    for(auto it = s.begin(); it != s.end(); it++) {
+        std::cout << *it << std::endl;
+    }
+
+    return 0;
+}
+
+
+// How do iterators describe ranges of items?
+
+// By the start() and end() methods which points out the end and beginning of for example, a vector
+
+// How do iterators look like for empty ranges? The same, but they're not going to compile;)
```

Comment:
The problem is it _is_ going to compile but your `start` an `end` iterators will point to the same location - one _past_ the end. And you're not allowed to dereference (`*it`) the end iterator. It will result in a out of bounds violation at runtime which maybe will crash, maybe will give you random memory - no guarantees. This is called: `undefined behavior`.